### ISSUE TITLE
Need a DAG edge remover/detector

### ISSUE THREAD

**Issue by godott** [Qiskit/qiskit-terra#1501]
1. Right now if there are two edges between two nodes(can happen between two CNOTs share two common qubits), dag.multi_graph.remove_edge(node1, nod2) can't specify which one to remove.
2. dag.multi_graph.has_edge(node1, node2) can only tell if there's an edge between two nodes, but can't tell if there's an edge between the two nodes on a particular qubit wire.

### What is the expected behavior?
1. Given two nodes we can remove any subset of edges between them.
2. Given two nodes we can tell if there is an edge on a particular qubit-wire.

**Comment by 1ucian0** [comment]
Our current DAG should not support `remove_edge` because the API operated through node operations. Being able to remove nodes allows to create "invalid" DAGs. I think #1768 should be reversed.

**Comment by 1ucian0** [comment]
Tagged as `wont fix`. Fell free to reopen it if think that I'm misunderstanding the issue.

**Comment by ajavadia** [comment]
@godott Luciano is right that the "DAG" should be a dataflow graph, showing the wires on which information flows from gate to gate. We need to create a separate "dependency graph" for commutations, but not modify the DAG in this way. A CNOT with one edge will be invalid in the DAG, but ok in the dependency graph.

@1ucian0 But I don't think we should remove the edge detector. That will not create invalid DAGs, and is needed for the commutation analysis to create that dependency graph.

**Comment by 1ucian0** [comment]
On Sat, Mar 9, 2019, 3:09 AM Ali Javadi-Abhari <notifications@github.com>
wrote:

> @1ucian0 <https://github.com/1ucian0> But I don't think we should remove
> the edge detector. That will not create invalid DAGs, and is needed for the
> commutation analysis to create that dependency graph.
>
Doesn't `successors()` cover that functionality?

>

**Comment by ajavadia** [comment]
No, that gives an iterator over successor nodes.
Here we want to ask "given CNOT2 is a successor of CNOT1, which wires (qubits) connect the two nodes?".. 
Actually maybe just `node['qargs']` can be used here. @godott would that work?

**Comment by godott** [comment]
@ajavadia I think `node['qargs']` cannot be used. One example is that two CNOTs both apply on q1 and q2 but separated by a 1-qubit gate on q1. If you ask gate successor, it will tell you  CNOT2 is the successor of CNOT1. Then If you ask `node['qargs']`, it will tell you that these two CNOTs both apply on q1 and q2. Then we might get the false conclusion that CNOT2 is a successor of CNOT1 both on q1 and q2, while it's only a successor on q2.

**Comment by godott** [comment]
Based on what @1ucian0 said, I agree that remover is not needed. I also agree with @ajavadia that detector is needed.

**Comment by 1ucian0** [comment]
If I understand correctly, `self.dag.has_edge(1, 2)` can be rewritten as `2 in self.dag.successors(1)`. Am I missing something?

**Comment by ajavadia** [comment]
ok i see.

So yeah maybe our API is lacking for this example you mention:
```
             ┌───┐
q_0: |0>──■──┤ Z ├──■──
        ┌─┴─┐└───┘┌─┴─┐
q_1: |0>┤ X ├─────┤ X ├
        └───┘     └───┘
```

Seems like you want to get successors **per wire** (say for example for CNOT1, the successor on q[0] is Z and the successor on q[1] is CNOT2). Not just that Z and CNOT2 are successors.

**Comment by godott** [comment]
Right. But if we don't need a DAG for commutativity anymore, then this functionality is not required either.

**Comment by maddy-tod** [comment]
The method introduced in #1975 (to fix #1502) provides an iterator over the gates on a given wire, so this would perform the functionality you described @ajavadia .

@1ucian0 you are correct, however by using `has_edge()` you can also specify if there is an edge on a given wire, which was what the original need was for (I don't know if this is still needed)
